projects/dendrites/subnetworks/utils.py

```diff
@@ -0,0 +1,383 @@
+
+def apply_target_transform(Y, offset=0):
+    """Transform label for continual learning problems"""
+    return [label + offset for label in Y]
+
+
+def add_task_embedding(X, task_num=1, units=2):
+    """Adds a task embedding to the input"""
+    assert task_num < units, \
+        "Task num are 0-indexed should be equal to max units-1"
+
+    num_rows = X.shape[0]
+    embedding = np.zeros((num_rows, units))
+    embedding[:, task_num] = 1
+    return np.hstack([X, embedding])
+
+
+def prep_dataset(datasets, embed_task=False, random_state=None,
+                 target_transform=False, target_transform_offset=3):
+    """Accepts an array of tuples (X, Y, name) or a single tuple
+    Return an array of tuples (Xs, Ys, name) or a single tuple
+    where Xs and Ys are also a tuple containing data for training and test"""
+
+    if type(datasets) != list:
+        datasets = [datasets]
+
+    ready_datasets = []
+    for idx, (X, Y, dataset_name) in enumerate(datasets):
+        if target_transform:
+            Y = apply_target_transform(Y, target_transform_offset * idx)
+        if embed_task:
+            X = add_task_embedding(X, task_num=idx)
+        Xs, Ys = split_data(X, Y, random_state=random_state)
+        ready_datasets.append((Xs, Ys, dataset_name))
+
+    if len(ready_datasets) == 1:
+        return ready_datasets[0]
+    return ready_datasets
+
+
+def load_datasets():
+    """Load Iris, Wine and Mixed datasets"""
+    # load datsets
+    iris = datasets.load_iris()
+    wine = datasets.load_wine()
+
+    # rebalance Wine dataset to have only 150 sample, same as iris
+    wine_indices_class0 = np.random.choice(range(59), size=51, replace=False)
+    wine_indices_class1 = np.random.choice(range(59, 59 + 71), size=51, replace=False)
+    wine_indices_class2 = np.array(range(59 + 71, 178))
+    wine_indices = np.sort(np.concatenate([
+        wine_indices_class0, wine_indices_class1, wine_indices_class2
+    ]))
+
+    # select 4 most relevant features, using feature selection methods:
+    # proline 12, od280/od315_of_diluted_wines 11
+    # flavanoids 6, color_intensity 9
+    wineX = wine.data[:, [6, 9, 11, 12]][wine_indices]
+
+    # scale both datasets from 0 to 1
+    irisX = MinMaxScaler().fit_transform(iris.data)
+    irisY = iris.target
+    wineX = MinMaxScaler().fit_transform(wineX)
+    wineY = wine.target[wine_indices]
+
+    # create mixed dataset
+    mixedX = np.concatenate([irisX, wineX])
+    mixedY = np.concatenate([irisY, apply_target_transform(wineY, 3)])
+
+    return (irisX, irisY, "Iris"), (wineX, wineY, "Wine"), (mixedX, mixedY, "Mixed")
+
+
+# ------ Training functions
+def evaluate(model, data, target, dataset_name=""):
+    pred = torch.argmax(model(torch.FloatTensor(data)), dim=1).numpy()
+    print(f"{dataset_name} Acc: {accuracy_score(target, pred):.4f}")
+
+def train(model, data, target, dataset_name, batch_size=1,
+          epochs=10, lr=1e-2, weight_decay=0, verbose=True):
```

Comment:
how is this training function different from the one in [model utils](https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/pytorch/model_utils.py#L37-L183), and why can't we re-use that?

Reply:
Not really different, but this is a small self-contained experiment. I needed the flexibility to modify the function as needed.